### PR TITLE
Add subroutes

### PR DESCRIPTION
**Pull request by sbatson5**
WIP PR for adding sub-routes per:
https://github.com/emberjs/guides/issues/1374

### CODE REVIEW

File: app/routes/rentals/show.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import Ember from 'ember';
+
+export default Ember.Route.extend({
+  model(params) {
+    return this.store.queryRecord('rental', { title: params.title })
+  },
+
+  setupController(controller, model) {
```
Comment (locks) [inline app/routes/rentals/show.js]: me no like, is this in the prose @toddjordan ?

Comment (sbatson5) [inline app/routes/rentals/show.js]: I thought it made the template more readable to access it as `rental` instead of `model` but I can revert it, nbd.  Is it considered bad practice to use `setupController` in this way?

Comment (locks) [inline app/routes/rentals/show.js]: @sbatson5 it's slightly controversial, but I personally think it's unnecessary and unexpected, since you expect the `model` property to be set with what's returned from the model hook.

Comment (locks) [inline app/routes/rentals/show.js]: in other words, there's some valid use cases for `setupController` but they're fewer than most people think they are ;)

File: app/templates/rentals/show.hbs [hunk 2]
```diff
@@ -0,0 +1 @@
+{{rental-listing rental=rental}}
```
Comment (toddjordan) [inline app/templates/rentals/show.hbs]: fyi, @kidgodzilla is working on providing markup/style for the details page.  Just showing the listing component is a great placeholder though.

**Comment by toddjordan** [comment]
Looks good :-)  I was thinking that you could use transitionTo from the index route to redirect users going to '/' to the new rentals route. http://emberjs.com/api/classes/Ember.Route.html#method_transitionTo

Also take a look at the last merge, we've added styling and changed the markup a bit.

File: app/templates/rentals/index.hbs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<ul>
```
Comment (toddjordan) [inline app/templates/rentals/index.hbs]: I'd expect this template to reflect the content that currently show in the homepage (using this template instead)

File: app/templates/rentals.hbs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<h2>Look up information for a specific rental</h2>
+
+{{outlet}}
+
+<footer>
```
Comment (toddjordan) [inline app/templates/rentals.hbs]: See the index route. we are no longer using a footer, but rather a top banner.  See the current index.hbs

File: tests/unit/routes/rentals/show-test.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { moduleFor, test } from 'ember-qunit';
```
Comment (toddjordan) [inline tests/unit/routes/rentals/show-test.js:1]: 👍 on the unit test

File: app/router.js [hunk 6]
```diff
@@ -8,6 +8,9 @@ const Router = Ember.Router.extend({
 Router.map(function() {
   this.route('about');
   this.route('contact');
+  this.route('rentals', function() {
+    this.route('show', { path: '/:title' });
```
Comment (toddjordan) [inline app/router.js]: So I've never had to put a variable with spaces into a route.  Will the router make this url friendly?  For example, if the title was "Beautiful House on the Beach" would the url end up being `/rentals/beautiful-house-on-the-beach` or would there be any code required to make this happen?   That's what I was getting at with the original suggestion to dasherize.

Comment (sbatson5) [inline app/router.js]: You would get: `/rentals/Beautiful%20house%20on%20the%20beach`.  The problem I ran into with `dasherize` is matching the title in the ember-cli-mirage mock.  We would have to add dasherize logic to the config file in order to match:
https://github.com/emberjs/super-rentals/pull/24/files#diff-04c022997e2228b6dc98593a3c6b7e87R58

Comment (toddjordan) [inline app/router.js]: Hmm, definitely wouldn't change it at the data level.  Was wondering if there was a way at the route level, but if its a non-common or obscure technique, I'm fine leaving out of the tutorial and going with the url encoded spaces.

Comment (acorncom) [inline app/router.js]: Could we add a `slug` field that would have this pre-built as part of the Mirage data?  It tends to fit better with the way it would generally be done.  Either that or use ids to help expose those and make it clearer what is going on ...

Comment (sbatson5) [inline app/router.js]: I like the slug approach as most API's would have one available. Should that be addressed in a separate PR?

Comment (acorncom) [inline app/router.js]: @sbatson5 Let's just add it to this one, makes it quite clear why we're adding it :-)

File: app/templates/rentals/index.hbs [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<ul class="results">
+  {{#each model as |rentalUnit|}}
+    {{rental-listing rental=rentalUnit goToShowRoute="goToShowRoute"}}
```
Comment (sbatson5) [inline app/templates/rentals/index.hbs]: Added an action to make the component clickable per recommendation here:
https://github.com/emberjs/guides/pull/1443#discussion_r64994911

Wondering if this is out of scope for a tutorial on sub-routes.  Not sure I want to encourage writing actions this way with closure actions and the route-action helper becoming more prominent. But I don't want to introduce a controller just for the sake of a closure action.

Anyone have thoughts on a faster way to do this that doesn't take much explaining for a beginner's tutorial?  Maybe just wrap component in a div and add the action to that?

Comment (toddjordan) [inline app/templates/rentals/index.hbs]: @sbatson5 You should just wrap rental listing in a {{link-to}} like this:

``` hbs
    {{#link-to "rentals.show" rentalUnit.slug}}
      {{rental-listing rental=rentalUnit}}
    {{/link-to}}
```

and remove all the "goToShowRoute" stuff

File: app/routes/rentals/show.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import Ember from 'ember';
+
+export default Ember.Route.extend({
+  model(params) {
+    return this.store.queryRecord('rental', { slug: params.slug });
+  },
+
+  setupController(controller, model) {
```
Comment (toddjordan) [inline app/routes/rentals/show.js]: My opinion on `setupController` is while I don't have anything against it, I think that its something extra that we don't need to introduce/explain in the beginner's tutorial.  I'd prefer to refer to the model as `model`, especially with the uncertain future for the controller.

**Comment by toddjordan** [comment]
I created a PR to your branch with what I was talking about: https://github.com/sbatson5/super-rentals/pull/1

File: app/templates/rentals/index.hbs (hunk 7 above)
Comment (sbatson5) [inline app/templates/rentals/index.hbs]: 👍 I was definitely overcomplicating it 😛

File: app/routes/rentals/show.js (hunk 8 above)
Comment (sbatson5) [inline app/routes/rentals/show.js]: Sounds good.  Will remove. Thanks!

File: app/templates/rentals/show.hbs [hunk 9]
```diff
@@ -0,0 +1,22 @@
+<div class="jumbo">
+  <a href="#" class="button right up-11">Edit</a>
```
Comment (sbatson5) [inline app/templates/rentals/show.hbs]: @toddjordan Do you want an `edit` subroute added to this tutorial?

Comment (toddjordan) [inline app/templates/rentals/show.hbs]: good question.  We've been recently discussing whether to add a form section to the tutorial, and edit would be the logical place to do it.  What I think is having the one subroute is fine for now, and when we add forms to the tutorial is when we would add the edit subroute.

File: tests/unit/routes/rentals/show-test.js (hunk 5 above)
Comment (sbatson5) [inline tests/unit/routes/rentals/show-test.js:1]: Reverted the unit test back since we're not doing the `setupController` hook.

Comment (toddjordan) [inline tests/unit/routes/rentals/show-test.js:1]: We should add a route test that verifies that the model is set to the selected rental.  The slightly funky thing is that in practice we would probably be using something like pretender or mirage to stub these calls, but we are using it anyways for the tutorial.

File: app/templates/rentals.hbs [hunk 10]
```diff
@@ -0,0 +1,13 @@
+<div class="jumbo">
+  <div class="right tomster"></div>
+  <h2>Welcome!</h2>
+  <p>
+    We hope you find exactly what you're looking for in a place to stay.
```
Comment (toddjordan) [inline app/templates/rentals.hbs:6]: We should do 1 of 2 things for this text:
1. Make the text more generic since it shows on both show and list
2. Make the text dynamic, to show different instructions based on the route.

File: tests/unit/routes/rentals/show-test.js (hunk 5 above)
Comment (sbatson5) [inline tests/unit/routes/rentals/show-test.js:1]: I meant to ask about that. I'm used to using pretender and acceptance tests to test that the model is being returned correctly. It seemed out of the scope of this tutorial, however because acceptance tests are a whole other beast.

Thoughts on how to properly test this? Using mirage in the app means that we are essentially just testing that `queryRecord` creates the proper `GET` request for us (which of course it does).

Comment (toddjordan) [inline tests/unit/routes/rentals/show-test.js:1]: An acceptance test section is currently being PR'd https://github.com/emberjs/guides/pull/1440
We may want to add a "show" test to the the acceptance tests either as part of or after this feature.

I think its fair to also test show route with a unit test, that the slug is being properly passed to the queryRecord call.  Should be fairly simple to add and will demonstrate route unit tests.

Comment (sbatson5) [inline tests/unit/routes/rentals/show-test.js:1]: Let me know if this revised unit test makes sense.

File: tests/unit/routes/rentals/show-test.js [hunk 11]
```diff
@@ -1,11 +1,14 @@
...
-moduleFor('route:rentals/show', 'Unit | Route | rentals/show', {
-  // Specify the other units that are required for this test.
-  // needs: ['controller:foo']
-});
+moduleFor('route:rentals/show', 'Unit | Route | rentals/show', {});
 
-test('it exists', function(assert) {
-  let route = this.subject();
-  assert.ok(route);
+test('queryRecord hits the API with the correct credentials', function(assert) {
```
Comment (toddjordan) [inline tests/unit/routes/rentals/show-test.js]: I've been starting my tests with "should", so that when run they end up reading like:
`Unit | Route | rentals/show should query for selected rental by slug`

File: app/templates/rentals/index.hbs [hunk 12]
```diff
@@ -0,0 +1,7 @@
+<ul class="results">
+  {{#each model as |rentalUnit|}}
+    {{#link-to "rentals.show" rentalUnit.slug}}
```
Comment (sbatson5) [inline app/templates/rentals/index.hbs]: My only hesitation with the `link-to` is that we don't get the benefit of expanding the component to show the larger image.

File: app/templates/rentals.hbs (hunk 10 above)
Comment (sbatson5) [inline app/templates/rentals.hbs:6]: In the vein of keeping this simple, generic text seems like the right solution.

File: app/templates/rentals/index.hbs (hunk 12 above)
Comment (toddjordan) [inline app/templates/rentals/index.hbs]: I agree we shouldn't lose that functionality.  One of us should do some research on if there's a way to supress/override the link to behavior with the image expansion.

File: tests/unit/routes/rentals/show-test.js (hunk 5 above)
Comment (toddjordan) [inline tests/unit/routes/rentals/show-test.js:1]: looks like this test just got reverted to the default...

Comment (sbatson5) [inline tests/unit/routes/rentals/show-test.js:1]: Odd.  Should be fixed now.

File: app/templates/rentals/index.hbs (hunk 12 above)
Comment (sbatson5) [inline app/templates/rentals/index.hbs]: I'm not finding an easy way to do this as we essentially have to make our component aware of our parent.  I had a solution with bubbling actions and preventing the default action when the target is the image, but that adds a lot of unnecessary complexity to something that is supposed to be simple.

The easiest solution may just be adding a link to the component saying "View Details" or something.

Comment (toddjordan) [inline app/templates/rentals/index.hbs]: OK.  We could link the title and within the listing component with a link-to, or we could create a toolbar above the listing with actions.  I think first options would be simplest and move us forward...

Comment (sbatson5) [inline app/templates/rentals/index.hbs]: Sorry for the delay in getting back to you on this.

The issue with turning the title into a link-to is that a `link-to`'s a tag is a block element whereas we need that h3 to be inline (otherwise it messes up all the styling).  It should only be 1 line of css in the add-on repo.  If we went with the nav option, we would have to add css anyway.

Alternatively, we could use an anchor and just add an inline style:
`<h3><a href="rental/show/{{rental.slug}}"style="display: inline">{{rental.title}}</a></h3>`

Comment (toddjordan) [inline app/templates/rentals/index.hbs]: We can totally PR the CSS addon repo with anything we need to make super-rentals right: https://github.com/toddjordan/ember-cli-tutorial-style

**Comment by sbatson5** [comment]
@toddjordan Thanks!  Added a PR: https://github.com/toddjordan/ember-cli-tutorial-style/pull/1

**Comment by toddjordan** [comment]
Cool.  I'll merge and re-publish the addon

**Comment by toddjordan** [comment]
@sbatson5 I updated the super-rentals repo with the updated version of the style addon with your update.  you can rebase it into this pr...

Also we'll need to squash commits before merge.

**Comment by toddjordan** [comment]
also, currently seems some of the tests are failing in this pr

**Comment by sbatson5** [comment]
Rebased and squashed commits. But not sure why Travis is failing. Have any insight @toddjordan ?

Also, I saw your PR I merged into this branch had changes to the `vendor/gmap.js`. Is that intended? Or should I remove it from this PR?

File: app/templates/rentals.hbs (hunk 10 above)
Comment (sbatson5) [inline app/templates/rentals.hbs:6]: Simply removed the second line. I think it looks fine unless anyone has a good suggestion on generic text.

**Comment by toddjordan** [comment]
I pulled down the latest from your branch and tests all pass (we're close!).  I did notice this though: 

image

Would prefer the text to sit between the picture and the map instead of below it.

File: app/templates/rentals/show.hbs (hunk 9 above)
Comment (toddjordan) [inline app/templates/rentals/show.hbs]: So go ahead and remove edit and we'll add it back when we introduce a forms section

**Comment by toddjordan** [comment]
I was able to correct the listing style issue by updating the ember-cli-tutorial-style addon to be more specific about an anchor style.  It seems adding the anchor to the listing for the subroute was picking up an unwanted style.

**Comment by toddjordan** [comment]
I updated the addon, all you'll need to do to fix is Update the ember-cli-tutorial-style version to 0.0.4!

I think it'll be then be good to go then and we can focus on getting the tutorial guide page out 👍

**Comment by sbatson5** [comment]
@toddjordan Thanks so much for catching that. I saw that styling last time but just thought that npm hadn't updated.  Forgot to go back and check it.  Will update.

**Comment by sbatson5** [comment]
@toddjordan I updated it to `0.0.4`, nombom'ed but am still seeing that odd styling.

**Comment by toddjordan** [comment]
Hmm I'll recheck tomorrow

**Comment by sbatson5** [comment]
I'm kind of stumped on this. When I look in `node_modules`, I can see all the css updates. But if I inspect the stylesheets in Chrome and Firefox, I am not seeing the css changes you or I added to the addon even after nombom'ing and running in incognito.  I feel there is something obvious I'm missing.

**Comment by toddjordan** [comment]
Sorry been on vacation and didn't get to it as planned. Should be able to
help out Sunday or Monday

**Comment by toddjordan** [comment]
@sbatson5 How did you update? The vendor css only gets updated when you run the blueprint, which comes from `ember install`, so your best course to update might be to remove the addon and re-run ember install, so it picks up the newest version and runs the blueprint

**Comment by sbatson5** [comment]
Ah, I was just doing `npm install`. I thought `ember serve` would update the vendor file. Good to know! It looks much better now.

**Comment by toddjordan** [comment]
Awesome.  I think this is good to go.  I'll hold off on merge until we merge the tutorial guides PR.  I'll look that that this week.

**Comment by toddjordan** [comment]
Created a new branch out of this one for remaining updates.  Closing in favor of: https://github.com/emberjs/super-rentals/pull/35
